STM32/ACTenChannel/Core/Src/ACTenChannel.c

```diff
@@ -0,0 +1,395 @@
+                        i, stmGetGpio(heaterPorts[i]), getPWMPinPercent(i));
+    }
+
+    writeUSB(buf, len);
+}
+
+static void userInput(const char *input)
+{
+    if (strncmp(input, "fan on", 6) == 0)
+    {
+        isFanForceOn = true;
+        stmSetGpio(fanCtrl, true);
+    }
+    else if (strncmp(input, "fan off", 7) == 0)
+    {
+        isFanForceOn = false;
+        stmSetGpio(fanCtrl, false);
+    }
+    else 
+    {
+        HALundefined(input);
+    }
+}
+
+static void GpioInit()
+{
+    const int NO_PORTS = AC_TEN_CH_NUM_PORTS;
+    static GPIO_TypeDef *const pinsBlk[] = { CTRL_0_GPIO_Port, CTRL_1_GPIO_Port, CTRL_2_GPIO_Port, CTRL_3_GPIO_Port, CTRL_4_GPIO_Port,
+                                             CTRL_5_GPIO_Port, CTRL_6_GPIO_Port, CTRL_7_GPIO_Port, CTRL_8_GPIO_Port, CTRL_9_GPIO_Port };
+    static const uint16_t pins[] = { CTRL_0_Pin, CTRL_1_Pin, CTRL_2_Pin, CTRL_3_Pin, CTRL_4_Pin,
+                                     CTRL_5_Pin, CTRL_6_Pin, CTRL_7_Pin, CTRL_8_Pin, CTRL_9_Pin };
+
+    for (int i = 0; i < NO_PORTS; i++)
+    {
+        stmGpioInit(&heaterPorts[i], pinsBlk[i], pins[i], STM_GPIO_OUTPUT);
+        heatCtrlAdd(&heaterPorts[i]);
+    }
+
+    stmGpioInit(&fanCtrl, FAN_CTRL_GPIO_Port, FAN_CTRL_Pin, STM_GPIO_OUTPUT);
+}
+
+static double ADCtoCurrent(double adc_val)
+{
+    static float CURRENT_SCALAR = 0.013138;
+    static float CURRENT_BIAS = -0.01;
+
+    return CURRENT_SCALAR * adc_val + CURRENT_BIAS;
+}
+
+static void printCurrentArray(int16_t *pData, int noOfChannels, int noOfSamples)
+{
+    // Make calibration static since this should be done only once.
+    static bool isCalibrationDone = false;
+    static int16_t current_calibration[ADC_CHANNELS];
+    static uint32_t port_close_time = 0;
+
+    /* If the USB port is not open, no messages should be printed. Also if the USB port has been 
+    ** closed for more than a timeout, everything should be turned off as a safety measure */
+    if (!isUsbPortOpen()) 
+    {
+        if (port_close_time == 0)
+        {
+            port_close_time = HAL_GetTick();
+        }
+        else if((HAL_GetTick() - port_close_time) >= USB_COMMS_TIMEOUT_MS)
+        {
+            allOff();
+        }
+        return;
+    }
+    port_close_time = 0;
+
+    /* If the version is incorrect, there is no point printing data or doing maths */
+    if (bsGetStatus() & BS_VERSION_ERROR_Msk)
+    {
+        USBnprintf("0x%08" PRIx32, bsGetStatus());
+        return;
+    }
+
+    if (!isCalibrationDone)
+    {
+        for (int i = 0; i < noOfChannels; i++)
+        {
+            // finding the average of each channel array to subtract from the readings
+            current_calibration[i] = -ADCMean(pData, i);
+        }
+        isCalibrationDone = true;
+    }
+
+    // Set bias for each channel.
+    for (int i = 0; i < noOfChannels; i++)
+    {
+        // Go from channel 1 since 0 is temperature.
```

Comment:
It seems that you do it for temperature as well now so you can remove this comment

Reply:
Yeah the comment does not make sense anymore. Previously, the temperature was on ADC channel 0, but it is now measured via the DS18B20. Anyway, I'll remove the comment.